docs/CLI.md

```diff
@@ -0,0 +1,158 @@
+# Dagger CLI
+
+> This is currently a throwaway markdown file for collaborating in a PR. Could
+> be turned into a doc but that's not the current priority.
+
+## CLI bird's eye view
+
+Here are all the commands for things you can do, with alternatives inline for
+side-by-side comparison.
+
+* run tests/checks
+    * TODO decide between `test` or `check`
+        * `test` is arguably more familiar and guessable, but not every check
+          will be a test suite. It could also be a linter, or whatever the user
+          wants to run.
+        * Maybe we should just go with `test` anyway, favoring familiarity over
+          technical correctness? Arguably non-test-sutie checks are just
+          integration tests. :grin:
+    * `dagger check` - run all checks
+    * `dagger check engine` - run a named check
+    * `dagger check engine -- -run Services` - run a named check, passing args
+    * `dagger test` - alternative
+    * `dagger test engine`
+    * `dagger test engine -- -run Services`
+* build artifacts
+    * `dagger build`
+* list available artifacts
+    * `dagger artifacts`
+* extend your environment with another environment
+    * `dagger use go` - universe environment
```

Comment:
Interested in exploring other options, but I was thinking more along the lines of Rust/Perl/PHP/a few other languages where `use` is used for loading modules/packages.

* https://doc.rust-lang.org/std/keyword.use.html
* https://perldoc.perl.org/functions/use
* https://www.php.net/manual/en/language.namespaces.importing.php

I've used it in SQL too, but always felt like that one was the oddball. :grin:

Reply:
I was thinking of it from the end user's perspective. They would run `dagger use go` to bootstrap their Go project, so without touching code they could do something like this:

```sh
$ dagger use go
$ dagger test
```

The default behavior is to merge it into the current environment, which may be initially empty/not exist. There could maybe be an optional namespace argument, which I imagine :tophat: would make use of.

Having said that, there are important details missing here:

* Where do I specify the base image that provides the `go` CLI? (Above assumes some sort of default I guess.)
* How do I go from this to modifying my local environment so that `dagger test` runs with `gotestsum` instead? (Assuming the `go` environment provides it.)
* If I `use` an environment but give it a namespace, how do I invoke its entrypoints?
* Do "merged" environments also merge entrypoints? (Above assumes yes.)

Reply:
Yeah good point.

Related thought: ideally we can support the workdir/env split while still not requiring you to pass an `--env` flag on every `dagger` call. I want my muscle memory to be just `dagger test` not `dagger -e ci test`. Given that, all `dagger` CLI commands should be able to discover the current environment even when run from your project workdir.

So `dagger use` should be able to detect when no environment is present and walk you through some first-time setup prompts, one of which is to ask where the environment should be bootstrapped, with strong hints that it is not the same as your current directory, and perhaps a warning if you try. It could also ask where the workdir of the project lives, with `./` as a default.

So the workflow could look something like this, for the super simple Progrock dev bootstrapping their side project:

```sh
$ dagger use go
It looks like you are starting a new dev environment.

Where would you like to put the environment code?
> ./ci

Where does your project code live?
> ./

Created ./ci/dagger.json        # environment metadata file
Created ./ci/dagger.lock        # environment dependencies
Created ./ci/dagger/api.gen.go  # generated environment client code

All done!
$ dagger test
```

Reply:
> ideally we can support the workdir/env split while still not requiring you to pass an `--env` flag on every `dagger` call. I want my muscle memory to be just `dagger test` not `dagger -e ci test`.

Yes, I agree. We must achieve this.

> Given that, all `dagger` CLI commands should be able to discover the current environment even when run from your project workdir.

Also agree.

> So `dagger use` should be able to detect when no environment is present and walk you through some first-time setup prompts, one of which is to ask where the environment should be bootstrapped, 

That feels like bringing back a regular embedded model.

> So the workflow could look something like this, for the super simple Progrock dev bootstrapping their side project:
> 
> ```shell
> $ dagger use go
> It looks like you are starting a new dev environment.
> 
> Where would you like to put the environment code?
> > ./ci
> 
> Where does your project code live?
> > ./
> 
> Created ./ci/dagger.json        # environment metadata file
> Created ./ci/dagger.lock        # environment dependencies
> Created ./ci/dagger/api.gen.go  # generated environment client code
> 
> All done!
> $ dagger test
> ```

This just feels like a more complicated version of the embedded model we use today. Might as well just commit the thing to `./ci` and be done with it.

Reply:
For something like Progrock I wouldn't want to maintain a separate repo for its dev environment, so in that example it's following the embedded model, yes. I had bootstrapping in mind anticipating a desire to switch hats at some point and maintain an environment in-repo. Embedded for convenience, but still shareable via `dagger use github.com/vito/progrock/ci`.

Discarding that for a moment, we could also have `dagger use go` just be shorthand for "use `--env go` for all future `dagger` commands". So the following would be equivalent:

```sh
$ dagger --env go test

$ dagger use go
$ dagger test
```

Reply:
> For something like Progrock I wouldn't want to maintain a separate repo for its dev environment

Can we dig into why?

Not saying you're wrong (it's the most familiar way by far). But if you're not going to use it, that's a good indication that nobody will. So really what we're discussing here is whether to pursue the "standalone env" model at all.

Reply:
Sure!

For hobby projects I care about keeping everything in one place to minimize mental overhead. I also don't want to pollute my GitHub account with even more repos for every silly idea before even knowing if it's a good idea. These are side projects, so I need to be able to quickly pop in when I find time, change as much as possible with however much pent up energy I have, commit, push, and context switch away from the project until next time.

I don't ever want to worry about compatibility. I just want to focus on keeping `main` working. I want to be able to refactor my code and CI infra together in one PR. Let's say I started with `package main` living in `./cmd/foo` but decide to move it to the root of the repo for easier `go install` UX. If my environment is in my repo, I can just move code around and change the Ci code from `go build ./cmd/foo` to `go build .`. If my environment lives in a separate repo, I have introduced a point in time where all past versions of the environment are no longer compatible with the project code, and vice versa.

To be clear, I'm not arguing against support for external environments. I think that's important, but I think don't think it's something most people have to worry about, and it comes with its own costs. I've used that pattern before with Concourse, where we had [separate CI repo + release pipeline](https://github.com/concourse/ci/blob/master/pipelines/release.yml) that took the Concourse repo as an input. It worked great, but we had a lot of trouble whenever there was a change that introduced compatibility issues between the pipeline template and the way the code had to be packaged and shipped.

Reply:
+1 to everything @vito said on the topic of embedded vs external, including the fact that even though the embedded model is important, external is also very important (especially since external environments are often re-usable environments, which is important for building the ecosystem).

I'll also add that if some top-hat is a monorepo-fanatic, then telling them they *have* to create a separate repo could almost be a dealbreaker, just on a philosophical level.

I still think the idea of defaulting to external repo but making it easy to define a `defaultContext` that points to your local repo is a nice+simple way of supporting both models seamlessly.

Reply:
What if your collection of environments was more like your dotfiles - a monolithic collection of all your environments, curated and maintained as a whole, rather than a parallel universe of repos?

It would change the conversation from: "where does this new software project live? new repo or subdir in an existing repo?" to: "where do I store my toolbox?"

Reply:
For example, here's where I used to maintain my dev environment (regardless of project): https://gist.github.com/shykes/7bf0704be14a0a9288d7dfe98617c208

Reply:
Makes sense - so I could maintain a single github.com/vito/vito-env and use it for all of my personal projects like Progrock, Bass, and Booklit.

That would help with the too-many-repos problem, but it doesn't fix the fundamental compatibility issue, and it seems like it makes it more likely to come up. Let's say I'm working on Progrock a lot and along the way I've made a ton of improvements and refactors to my environment code, and those changes break Bass and Booklit's usage. Ideally the two don't break on `main` because they're pinned to a particular version of the environment, but I would still have to deal with that breakage when I bump the environment dependency.

The major benefit of sharing an environment is that I don't have to maintain similar code across all my projects, but I might prefer that pain if there's not much code to begin with. It feels a whole lot like judging when to refactor. Sometimes a bit of duplicate code is better than DRYing everything up because it keeps all the doors open in the future and keeps other things decoupled.

Reply:
Could there be a potential best practice that says something like this:

- If it defines app-specific dependencies, *don't encode it in the environment*: use your app's platform-specific package managers instead: `package.json`, `requirements.txt`, `nix-flakes-something-something`, `Dockerfile`...
- If it orchestrates *how* and *when* to install dependencies, encode it in the environment

In other words, we could use the Dagger/not-dagger boundary to expose a clean app/platform boundary.

Reply:
As a guy wearing the ops hat in a web agency, where you have several projects that share the same framework and deployment strategy, I had it kind of split. Meaning, a lot of duplication in each project for the dev environment (for team members), but a separate repo with config reuse just for deploying into production (privileged). I looked to dagger to help bridge this gap.

Now lets say I have 3 solo projects, each using different languages I'm not fully sure if I want to manage dagger in a separate repo. I wouldn't create a separate repo for each project, just a single one for all, but like @vito says, when I commit to get something done in a project, I prefer to deal with the code and dagger environment in a single repo and there's not much benefit to put in a separate repo if there's no reusability needs. 

In addition to that, I want to be able to run from working changes, not necessarily a committed change in a repo, to avoid the "fix deployment" kind of commits.

Having said that, I do think it's powerful to easily support both models. Let's say I have several Django projects.

First, I’d create a general “django” environment with all sorts of conventions. How would I test project “caloura”?
```console
~/projs/caloura$ dagger --env github.com/helderco/ci/django test
```

To avoid repeating the `--env` in commands, I can add the static dagger config to this project:
```console
~/projs/caloura$ dagger use github.com/helderco/ci/django
~/projs/caloura$ dagger test
```

But doing it outside, could also be possible (e.g., repo of project specific environments):

```console
~/projs/infra/caloura$ dagger use github.com/helderco/ci/django
~/projs/infra/caloura$ # "root": "github.com/helderco/caloura"
~/projs/infra/caloura$ dagger test
```

And to run against working changes:

```console
~/projs/infra/caloura$ dagger --root ~/proj/caloura test
```

So basically the static file can be a way to set defaults for cli arguments that can be overridden by specifying directly. A bit like a `compose.yml` file to `docker run`. As defaults go, the shortest command is to make the project directory the working dir (without a static file in the project).

The changes to my project above are independent from the dagger environment, but that’s not always the case. I’ve always needed to have some project specific configurations. But that could perhaps be encoded either in the project or the separate “infra” repo in a way that the environment code can read it and handle appropriately.

Another issue is that over time, these conventions (e.g., `ci/django` environment) change, but pinning versions in a `dagger.lock` prevents breakage.

*Note: I wrote this earlier today but now realized I didn’t submit.*

Reply:
> I still think the idea of defaulting to external repo but making it easy to define a `defaultContext` that points to your local repo is a nice+simple way of supporting both models seamlessly.

Yeah, the way I'm thinking about it makes it seamless to support both models.